Applications/LoRATrainingExample/ContentView.swift

```diff
@@ -0,0 +1,284 @@
+                    }
+                }
+                .padding()
+                .frame(maxWidth: .infinity)
+            }
+        }
+        .padding()
+    }
+
+    func start() {
+        Task {
+            await evaluator.start()
+        }
+    }
+
+    func evaluate() {
+        Task {
+            await evaluator.evaluate(prompt: prompt)
+        }
+    }
+}
+
+/// Progress reporting with a title.
+struct Progress: Equatable {
+    let title: String
+    let current: Double?
+    let limit: Double?
+}
+
+@Observable
+class LoRAEvaluator {
+
+    enum State {
+        case idle
+        case training
+        case evaluate
+        case failed(String)
+    }
+
+    enum ModelState {
+        case idle
+        case loaded(LLMModel, Tokenizer)
+    }
+
+    var state = State.idle
+    var progress: Progress?
+
+    var output = ""
+
+    private let modelConfiguration = ModelConfiguration.mistral7B4bit
+    private var model: ModelState = .idle
+
+    private let loraLayers = 4
+    private let learningRate: Float = 1e-5
+    private let parameters = LoRATrain.Parameters(batchSize: 1, iterations: 200)
+
+    private let generateParameters = GenerateParameters(temperature: 0.6, topP: 0.9)
+    private let evaluateShowEvery = 8
+    private let maxTokens = 200
+
+    private func loadModel() async throws -> (LLMModel, Tokenizer) {
+        switch self.model {
+        case .idle:
+            let name = modelConfiguration.name
+            await MainActor.run {
+                progress = .init(title: "Loading \(name)", current: 0, limit: 1)
+            }
+
+            let (model, tokenizer) = try await LLM.load(configuration: modelConfiguration) {
+                progress in
+                if progress.fractionCompleted < 1.0 {
+                    DispatchQueue.main.sync {
+                        self.progress = .init(
+                            title: "Download \(name)", current: progress.fractionCompleted,
+                            limit: 1.0)
+                    }
+                }
+            }
+            eval(model)
+            self.model = .loaded(model, tokenizer)
+            return (model, tokenizer)
+
+        case .loaded(let model, let tokenizer):
+            return (model, tokenizer)
+        }
+    }
+
+    private func loadLoRAData(name: String) throws -> [String]? {
+        if let url = Bundle.main.url(forResource: name, withExtension: "jsonl") {
+            return try LLM.loadLoRAData(url: url)
+        }
+        return nil
+    }
+
+    func start() async {
+        do {
+            try await startInner()
+        } catch {
+            self.state = .failed("Failed: \(error)")
+        }
+    }
+
+    private func startInner() async throws {
+        // setup
+        GPU.set(cacheLimit: 32 * 1024 * 1024)
+        await MainActor.run {
+            output = ""
+            state = .training
+        }
+
+        // load the model
+        let (model, tokenizer) = try await loadModel()
```

Comment:
@awni this is an example of the load / train / test loop.  At least half the code is just updating the UI :-)